pkg/controller/pkg/revision/imageback_test.go

```diff
@@ -0,0 +1,178 @@
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/random"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/pkg/errors"
+
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+	"github.com/crossplane/crossplane-runtime/pkg/test"
+	"github.com/crossplane/crossplane/pkg/xpkg"
+)
+
+type MockFetcher struct {
+	MockFetch func() (v1.Image, error)
+}
+
+func (m *MockFetcher) Fetch(ctx context.Context, ref name.Reference, secrets []string) (v1.Image, error) {
+	return m.MockFetch()
+}
+
+func TestImageBackend(t *testing.T) {
+	errBoom := errors.New("boom")
+	randLayer, _ := random.Layer(int64(1000), types.DockerLayer)
+	randImg, _ := mutate.AppendLayers(empty.Image, randLayer)
+
+	streamCont := "somestreamofyaml"
+	tarBuf := new(bytes.Buffer)
+	tw := tar.NewWriter(tarBuf)
+	hdr := &tar.Header{
+		Name: xpkg.StreamFile,
+		Mode: int64(xpkg.StreamFileMode),
+		Size: int64(len(streamCont)),
+	}
+	_ = tw.WriteHeader(hdr)
+	_, _ = io.Copy(tw, strings.NewReader(streamCont))
+	_ = tw.Close()
+	packLayer, _ := tarball.LayerFromReader(tarBuf)
+	packImg, _ := mutate.AppendLayers(empty.Image, packLayer)
+
+	type args struct {
+		c    xpkg.Cache
+		f    Fetcher
+		opts []parser.BackendOption
+	}
+
+	cases := map[string]struct {
+		reason string
+		args   args
+		want   error
+	}{
+		"ErrBadReference": {
+			reason: "Should return error if package tag is not a valid image reference.",
+			args: args{
+				opts: []parser.BackendOption{Package(":test")},
+			},
+			want: errors.Wrap(name.NewErrBadName("could not parse reference: :test"), errBadReference),
+		},
+		"ErrFetchedBadPackage": {
+			reason: "Should return error if image with contents does not have package.yaml.",
+			args: args{
+				c: xpkg.NewNopCache(),
+				f: &MockFetcher{
+					MockFetch: func() (v1.Image, error) {
+						return randImg, nil
+					},
+				},
+				opts: []parser.BackendOption{Package("test/test:latest")},
+			},
+			want: errors.Wrap(&os.PathError{Op: "open", Path: xpkg.StreamFile, Err: syscall.ENOENT}, errOpenPackageStream),
+		},
+		// TODO(hasheddan): re-enable after https://github.com/spf13/afero/pull/268 is merged.
+		// "ErrEmptyImage": {
+		// 	reason: "Should return error if image is empty.",
+		// 	args: args{
+		// 		c: xpkg.NewNopCache(),
+		// 		f: &MockFetcher{
+		// 			MockFetch: func() (v1.Image, error) {
+		// 				return empty.Image, nil
+		// 			},
+		// 		},
+		// 		opts: []parser.BackendOption{Package("test/test:latest")},
+		// 	},
+		// 	want: errors.Wrap(&os.PathError{Op: "open", Path: "package.yaml", Err: syscall.ENOENT}, errOpenPackageStream),
+		// },
```

Comment:
This test causes a panic because of the way the afero is handling tarfs initialization when an empty tar reader is passed. I have opened https://github.com/spf13/afero/pull/268 to fix. In practice, this would certainly be an edge case (i.e. totally empty package), but we obviously don't opportunity for panics in our code. I will work with afero maintainers to get this addressed prior to Crossplane release, or introduce a check and error early to avoid this execution branch.